### PR TITLE
add on-the-fly frequency encoder

### PR DESCRIPTION
**Pull request by guolinke**
Fixes and improves #71

This is like the combination of count and CatBoost. The frequency is calculated on-the-fly.
This is used for the chronological data.

### CODE REVIEW

**Comment by guolinke** [comment]
Although this is an unsupervised encoding, but it uses different encodings for the training set and test set. So the label is used for distinguishing the training set.
